src/app/auth/login.ts

```diff
@@ -113,6 +135,136 @@ export const loginWithPhoneCookie = async ({
+  //         * can a user manually sweep funds to new account then click a btn in the mobile app to delete device account?
+  //               // if device account bal is 0 and user clicks nuke account then delete kratos and mongo entries with
+  //                      mutation nukeDeviceAccountMutation ?
+  //
+  // Scenario 2) Happy Path - User logs in with device jwt, no phone account exists, upgrade device to phone account
+  //             a. update kratos => update schema to phone_no_password_v0, remove device trait
+  //             b. mongo (user) => update user collection and remove device field, add phone
+  //             c. mongo (account) => update account to level 1
+  //
+  // Scenario 3) Unhappy path - User sells phone and forgets to sweep funds to phone account
+  //             Option 1 - too bad, we can't help them
+  //             Option 2 - create some kind of recovery code process?
+
+  // Scenario 1 - does phone account already exist?
+  const phoneAccount = await UsersRepository().findByPhone(phone)
+  let hasPhoneAccount = true
+  if (phoneAccount instanceof CouldNotFindUserFromPhoneError) {
+    hasPhoneAccount = false
+  } else if (phoneAccount instanceof Error) {
+    return phoneAccount
+  } else if (hasPhoneAccount) {
+    // is there still txns left over on the device account?
+    const deviceWallets = await WalletsRepository().listByAccountId(account.id)
+    if (deviceWallets instanceof Error) return deviceWallets
+    const ledger = LedgerService()
+    let deviceAccountHasBalance = false
+    for (const wallet of deviceWallets) {
+      const balance = await ledger.getWalletBalance(wallet.id)
+      if (balance instanceof Error) return balance
+      if (balance > 0) {
+        deviceAccountHasBalance = true
+      }
+    }
+    if (deviceAccountHasBalance) {
+      // Scenario 1.5 - has txns on device account but phone account exists
+      return new PhoneAccountAlreadyExistsNeedToSweepFundsError()
+    } else {
+      // Scenario 1 - no txns on device account but phone account exists
+      return new PhoneAccountAlreadyExistsError()
+    }
+  }
+
+  // Scenario 2 - Happy Path
+  // a. create kratos account
+  // b. and c. migrate account/user collection in mongo via kratos/registration webhook
+  const kratosToken = await AuthWithDeviceAccountService().upgradeToPhoneSchema({
+    phone,
+    deviceId: account.kratosUserId,
+  })
+  if (kratosToken instanceof Error) return kratosToken
+  return kratosToken
+}
+
+export const loginWithDevice = async ({
+  jwt,
+  ip,
+}: {
+  jwt: string
+  ip: IpAddress
+}): Promise<SessionToken | ApplicationError> => {
+  {
+    const limitOk = await checkFailedLoginAttemptPerIpLimits(ip)
+    if (limitOk instanceof Error) return limitOk
+  }
+
+  const verifiedJwt = await verifyJwt(jwt)
+  if (verifiedJwt instanceof Error) {
+    await rewardFailedLoginAttemptPerIpLimits(ip)
+    return verifiedJwt
+  }
+  const deviceId = verifiedJwt.sub as UserId
```

Comment:
`verifiedJwt.sub` can be `undefined` based on the type information. we should do a check on it.